File: aermet/core/aermet_runner.py
"""
AERMET control file generation and execution.

Generates control files for all 3 AERMET stages:
  Stage 1: Extract & QA (surface + upper air data)
  Stage 2: Merge
  Stage 3: METPREP (boundary layer parameters → .SFC + .PFL output)

Executes AERMET.exe via subprocess and parses message files.
"""

import os
import subprocess

_BIN_DIR = os.path.join(os.path.dirname(__file__), '..', 'bin')


def _aermet_exe():
    """Locate AERMET executable."""
    exe = os.path.join(_BIN_DIR, 'aermet.exe')
    if os.path.exists(exe):
        return exe
    # Try without .exe for non-Windows
    exe2 = os.path.join(_BIN_DIR, 'aermet')
    if os.path.exists(exe2):
        return exe2
    return None


# ---------------------------------------------------------------------------
# Control file generation
# ---------------------------------------------------------------------------

def generate_stage1(work_dir, year, surface_file, upper_air_file,
                    sf_station_id, sf_lat, sf_lon,
                    ua_station_id, ua_lat, ua_lon,
                    start_month=1, end_month=12,
                    onemin_file=None):
    """
    Generate AERMET Stage 1 control file (EXTRACT + QA).

    Returns path to control file.
    """
    prefix = f's1_{year}'
    ctrl_path = os.path.join(work_dir, f'{prefix}.inp')

    start_date = f'{year}/01/01'
    if start_month != 1:
        start_date = f'{year}/{start_month:02d}/01'
    end_date = f'{year}/12/31'
    if end_month != 12:
        # Last day of month
        import calendar
        last_day = calendar.monthrange(year, end_month)[1]
        end_date = f'{year}/{end_month:02d}/{last_day}'

    # Build SURFACE keywords
    oneminute_line = ''
    if onemin_file:
        oneminute_line = f'   ONEMINUTE   {onemin_file}'

    ctrl = f"""JOB
   MESSAGES   {prefix}.msg
   REPORT     {prefix}.rpt

UPPERAIR
   DATA       {upper_air_file}  IGRA
   EXTRACT    {prefix}_ua.ext
   AUDIT      UATT UAWS UALR
   QAOUT      {prefix}_ua.qa
   XDATES     {start_date}  TO  {end_date}
   LOCATION   {ua_station_id}  {ua_lat:.4f}  {ua_lon:.4f}

SURFACE
   DATA       {surface_file}  ISHD
   EXTRACT    {prefix}_sf.ext
   AUDIT      SLVP PRES CLHT TSKC PWTH APTS
   QAOUT      {prefix}_sf.qa
   XDATES     {start_date}  TO  {end_date}
   LOCATION   {sf_station_id}  {sf_lat:.4f}  {sf_lon:.4f}
{oneminute_line}
"""

    with open(ctrl_path, 'w') as f:
        f.write(ctrl)
    return ctrl_path


def generate_stage2(work_dir, year, onemin_file=None):
    """Generate AERMET Stage 2 control file (MERGE)."""
    prefix = f's2_{year}'
    s1_prefix = f's1_{year}'
    ctrl_path = os.path.join(work_dir, f'{prefix}.inp')

    oneminute_line = ''
    if onemin_file:
        oneminute_line = f'   ONEMINUTE   {onemin_file}'

    ctrl = f"""JOB
   MESSAGES   {prefix}.msg
   REPORT     {prefix}.rpt

UPPERAIR
   QAOUT      {s1_prefix}_ua.qa

SURFACE
   QAOUT      {s1_prefix}_sf.qa
{oneminute_line}

MERGE
   OUTPUT     {prefix}_merge.out
   XDATES     {year}/01/01  TO  {year}/12/31
"""

    with open(ctrl_path, 'w') as f:
        f.write(ctrl)
    return ctrl_path


def generate_stage3(work_dir, year, sf_station_id, sf_lat, sf_lon,
                    ua_station_id, ua_lat, ua_lon,
                    aersurface_file=None, anemometer_height=10.0,
                    manual_surface_chars=None):
    """
    Generate AERMET Stage 3 control file (METPREP).

    Args:
        aersurface_file: Path to AERSURFACE output file (US sites)
        manual_surface_chars: Dict with albedo, bowen, roughness for non-US sites
    """
    prefix = f's3_{year}'
    s2_prefix = f's2_{year}'
    ctrl_path = os.path.join(work_dir, f'{prefix}.inp')

    sfc_out = os.path.join('output', f'{year}.SFC')
    pfl_out = os.path.join('output', f'{year}.PFL')

    # Surface characteristics section
    surf_chars = ''
    if aersurface_file:
        surf_chars = f'   AERSURF    {aersurface_file}'
    elif manual_surface_chars:
        # Manual entry: single sector, all seasons
        albedo = manual_surface_chars.get('albedo', 0.18)
        bowen = manual_surface_chars.get('bowen', 1.0)
        roughness = manual_surface_chars.get('roughness', 0.15)
        surf_chars = f"""   FREQ_SECT  ANNUAL  1
   SECTOR     1   0  360
   SITE_CHAR  1  1  {albedo}  {bowen}  {roughness}
   SITE_CHAR  1  2  {albedo}  {bowen}  {roughness}
   SITE_CHAR  1  3  {albedo}  {bowen}  {roughness}
   SITE_CHAR  1  4  {albedo}  {bowen}  {roughness}"""
    else:
        # Defaults
        surf_chars = """   FREQ_SECT  ANNUAL  1
   SECTOR     1   0  360
   SITE_CHAR  1  1  0.18  1.0  0.15
   SITE_CHAR  1  2  0.18  1.0  0.15
   SITE_CHAR  1  3  0.18  1.0  0.15
   SITE_CHAR  1  4  0.18  1.0  0.15"""

    ctrl = f"""JOB
   MESSAGES   {prefix}.msg
   REPORT     {prefix}.rpt

METPREP
   DATA       {s2_prefix}_merge.out
   XDATES     {year}/01/01  TO  {year}/12/31
   OUTPUT     {sfc_out}
   PROFILE    {pfl_out}
   SURFDATA   {sf_station_id}  {year}
   UAIRDATA   {ua_station_id}  {year}
   SURFLOCN   {sf_lat:.4f}  {sf_lon:.4f}
   UAIRLOCN   {ua_lat:.4f}  {ua_lon:.4f}
   NWS_HGT    {anemometer_height}
   METHOD     REFLEVEL  SUBNWS
{surf_chars}
"""

    with open(ctrl_path, 'w') as f:
        f.write(ctrl)
    return ctrl_path


# ---------------------------------------------------------------------------
# Execution
# ---------------------------------------------------------------------------

def run_aermet(ctrl_path, work_dir, progress_cb=None):
    """
    Execute AERMET with a given control file.

    Returns dict with success, return_code, messages, warnings, errors.
    """
    exe = _aermet_exe()
    if not exe:
        return {
            'success': False,
            'return_code': -1,
            'error': 'AERMET executable not found. Place aermet.exe in aermet/bin/',
            'messages': [],
            'warnings': [],
            'errors': ['AERMET executable not found'],
        }

    ctrl_filename = os.path.basename(ctrl_path)

    if progress_cb:
        progress_cb('running', f'Running AERMET with {ctrl_filename}...')

    try:
        result = subprocess.run(
            [exe],
            input=ctrl_filename + '\n',
            capture_output=True,
            text=True,
            cwd=work_dir,
            timeout=600,
        )

        # Parse the message file
        msg_file = ctrl_filename.replace('.inp', '.msg')
        msg_path = os.path.join(work_dir, msg_file)
        messages, warnings, errors = _parse_message_file(msg_path)

        success = result.returncode == 0 and len(errors) == 0

        if progress_cb:
            status = 'done' if success else 'error'
            progress_cb(status, f'AERMET {ctrl_filename}: {"OK" if success else "errors found"}')

        return {
            'success': success,
            'return_code': result.returncode,
            'stdout': result.stdout,
            'stderr': result.stderr,
            'messages': messages,
            'warnings': warnings,
            'errors': errors,
        }
    except FileNotFoundError:
        return {
            'success': False,
            'return_code': -1,
            'error': 'AERMET executable not found or not runnable',
            'messages': [],
            'warnings': [],
            'errors': ['AERMET executable not found'],
        }
    except subprocess.TimeoutExpired:
        return {
            'success': False,
            'return_code': -1,
            'error': 'AERMET timed out after 10 minutes',
            'messages': [],
            'warnings': [],
            'errors': ['AERMET execution timed out'],
        }


def run_aermet_all_stages(work_dir, year, surface_file, upper_air_file,
                           sf_station_id, sf_lat, sf_lon,
                           ua_station_id, ua_lat, ua_lon,
                           aersurface_file=None, onemin_file=None,
                           anemometer_height=10.0,
                           manual_surface_chars=None,
                           progress_cb=None):
    """
    Run all 3 AERMET stages for a single year.
    Returns dict with results from each stage and output file paths.
    """
    results = {}

    # Stage 1
    if progress_cb:
        progress_cb('stage1', f'AERMET Stage 1 ({year}): Extract & QA...')
    ctrl1 = generate_stage1(
        work_dir, year, surface_file, upper_air_file,
        sf_station_id, sf_lat, sf_lon,
        ua_station_id, ua_lat, ua_lon,
        onemin_file=onemin_file,
    )
    results['stage1'] = run_aermet(ctrl1, work_dir, progress_cb)
    if not results['stage1']['success']:
        return {'success': False, 'stage': 1, 'results': results}

    # Stage 2
    if progress_cb:
        progress_cb('stage2', f'AERMET Stage 2 ({year}): Merge...')
    ctrl2 = generate_stage2(work_dir, year, onemin_file=onemin_file)
    results['stage2'] = run_aermet(ctrl2, work_dir, progress_cb)
    if not results['stage2']['success']:
        return {'success': False, 'stage': 2, 'results': results}

    # Stage 3
    if progress_cb:
        progress_cb('stage3', f'AERMET Stage 3 ({year}): METPREP...')
    ctrl3 = generate_stage3(
        work_dir, year, sf_station_id, sf_lat, sf_lon,
        ua_station_id, ua_lat, ua_lon,
        aersurface_file=aersurface_file,
        anemometer_height=anemometer_height,
        manual_surface_chars=manual_surface_chars,
    )
    results['stage3'] = run_aermet(ctrl3, work_dir, progress_cb)

    sfc_path = os.path.join(work_dir, 'output', f'{year}.SFC')
    pfl_path = os.path.join(work_dir, 'output', f'{year}.PFL')

    return {
        'success': results['stage3']['success'],
        'results': results,
        'sfc_path': sfc_path,
        'pfl_path': pfl_path,
    }


# ---------------------------------------------------------------------------
# Message file parsing
# ---------------------------------------------------------------------------

def _parse_message_file(msg_path):
    """Parse an AERMET message file for messages, warnings, and errors."""
    messages = []
    warnings = []
    errors = []

    if not os.path.exists(msg_path):
        return messages, warnings, errors

    with open(msg_path, 'r', encoding='utf-8', errors='replace') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            if '***' in line and 'ERROR' in line.upper():
                errors.append(line)
            elif '***' in line and 'WARNING' in line.upper():
                warnings.append(line)
            elif line.startswith('***'):
                messages.append(line)
            # Also catch "W" and "E" prefixed messages
            elif line.startswith('E ') or line.startswith('E\t'):
                errors.append(line)
            elif line.startswith('W ') or line.startswith('W\t'):
                warnings.append(line)

    return messages, warnings, errors


def combine_sfc_files(sfc_files, output_path):
    """Combine multiple per-year .SFC files into one."""
    with open(output_path, 'w') as out:
        for i, path in enumerate(sorted(sfc_files)):
            with open(path, 'r') as f:
                for line in f:
                    out.write(line)
    return output_path


def combine_pfl_files(pfl_files, output_path):
    """Combine multiple per-year .PFL files into one."""
    with open(output_path, 'w') as out:
        for path in sorted(pfl_files):
            with open(path, 'r') as f:
                for line in f:
                    out.write(line)
    return output_path

File: aermet-tool/src/aermet_runner.rs
/// AERMET control file generation and execution.
///
/// Generates control files for all 3 AERMET stages and runs aermet.exe.
use serde::Serialize;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

fn aermet_exe() -> Option<PathBuf> {
    let exe_dir = std::env::current_exe()
        .ok()
        .and_then(|p| p.parent().map(|p| p.to_path_buf()))
        .unwrap_or_else(|| PathBuf::from("."));
    let exe = exe_dir.join("bin").join("aermet.exe");
    if exe.exists() {
        return Some(exe);
    }
    let exe2 = exe_dir.join("bin").join("aermet");
    if exe2.exists() {
        return Some(exe2);
    }
    None
}

// ---------------------------------------------------------------------------
// Control file generation
// ---------------------------------------------------------------------------

pub fn generate_stage1(
    work_dir: &Path,
    year: i32,
    surface_file: &str,
    upper_air_file: &str,
    sf_station_id: &str,
    sf_lat: f64,
    sf_lon: f64,
    ua_station_id: &str,
    ua_lat: f64,
    ua_lon: f64,
    onemin_file: Option<&str>,
) -> Result<PathBuf, String> {
    let prefix = format!("s1_{year}");
    let ctrl_path = work_dir.join(format!("{prefix}.inp"));
    let start_date = format!("{year}/01/01");
    let end_date = format!("{year}/12/31");

    let oneminute_line = onemin_file
        .map(|f| format!("   ONEMINUTE   {f}"))
        .unwrap_or_default();

    let ctrl = format!(
        "JOB\n\
         \x20  MESSAGES   {prefix}.msg\n\
         \x20  REPORT     {prefix}.rpt\n\
         \n\
         UPPERAIR\n\
         \x20  DATA       {upper_air_file}  IGRA\n\
         \x20  EXTRACT    {prefix}_ua.ext\n\
         \x20  AUDIT      UATT UAWS UALR\n\
         \x20  QAOUT      {prefix}_ua.qa\n\
         \x20  XDATES     {start_date}  TO  {end_date}\n\
         \x20  LOCATION   {ua_station_id}  {ua_lat:.4}  {ua_lon:.4}\n\
         \n\
         SURFACE\n\
         \x20  DATA       {surface_file}  ISHD\n\
         \x20  EXTRACT    {prefix}_sf.ext\n\
         \x20  AUDIT      SLVP PRES CLHT TSKC PWTH APTS\n\
         \x20  QAOUT      {prefix}_sf.qa\n\
         \x20  XDATES     {start_date}  TO  {end_date}\n\
         \x20  LOCATION   {sf_station_id}  {sf_lat:.4}  {sf_lon:.4}\n\
         {oneminute_line}\n"
    );

    fs::write(&ctrl_path, ctrl).map_err(|e| format!("Write stage1: {e}"))?;
    Ok(ctrl_path)
}

pub fn generate_stage2(
    work_dir: &Path,
    year: i32,
    onemin_file: Option<&str>,
) -> Result<PathBuf, String> {
    let prefix = format!("s2_{year}");
    let s1_prefix = format!("s1_{year}");
    let ctrl_path = work_dir.join(format!("{prefix}.inp"));

    let oneminute_line = onemin_file
        .map(|f| format!("   ONEMINUTE   {f}"))
        .unwrap_or_default();

    let ctrl = format!(
        "JOB\n\
         \x20  MESSAGES   {prefix}.msg\n\
         \x20  REPORT     {prefix}.rpt\n\
         \n\
         UPPERAIR\n\
         \x20  QAOUT      {s1_prefix}_ua.qa\n\
         \n\
         SURFACE\n\
         \x20  QAOUT      {s1_prefix}_sf.qa\n\
         {oneminute_line}\n\
         \n\
         MERGE\n\
         \x20  OUTPUT     {prefix}_merge.out\n\
         \x20  XDATES     {year}/01/01  TO  {year}/12/31\n"
    );

    fs::write(&ctrl_path, ctrl).map_err(|e| format!("Write stage2: {e}"))?;
    Ok(ctrl_path)
}

#[derive(Debug, Clone, Serialize, serde::Deserialize)]
pub struct ManualSurfaceChars {
    pub albedo: f64,
    pub bowen: f64,
    pub roughness: f64,
}

pub fn generate_stage3(
    work_dir: &Path,
    year: i32,
    sf_station_id: &str,
    sf_lat: f64,
    sf_lon: f64,
    ua_station_id: &str,
    ua_lat: f64,
    ua_lon: f64,
    aersurface_file: Option<&str>,
    anemometer_height: f64,
    manual_surface_chars: Option<&ManualSurfaceChars>,
) -> Result<PathBuf, String> {
    let prefix = format!("s3_{year}");
    let s2_prefix = format!("s2_{year}");
    let ctrl_path = work_dir.join(format!("{prefix}.inp"));
    let sfc_out = format!("output/{year}.SFC");
    let pfl_out = format!("output/{year}.PFL");

    let surf_chars = if let Some(asf) = aersurface_file {
        format!("   AERSURF    {asf}")
    } else if let Some(mc) = manual_surface_chars {
        format!(
            "   FREQ_SECT  ANNUAL  1\n\
             \x20  SECTOR     1   0  360\n\
             \x20  SITE_CHAR  1  1  {}  {}  {}\n\
             \x20  SITE_CHAR  1  2  {}  {}  {}\n\
             \x20  SITE_CHAR  1  3  {}  {}  {}\n\
             \x20  SITE_CHAR  1  4  {}  {}  {}",
            mc.albedo, mc.bowen, mc.roughness,
            mc.albedo, mc.bowen, mc.roughness,
            mc.albedo, mc.bowen, mc.roughness,
            mc.albedo, mc.bowen, mc.roughness,
        )
    } else {
        "   FREQ_SECT  ANNUAL  1\n\
         \x20  SECTOR     1   0  360\n\
         \x20  SITE_CHAR  1  1  0.18  1.0  0.15\n\
         \x20  SITE_CHAR  1  2  0.18  1.0  0.15\n\
         \x20  SITE_CHAR  1  3  0.18  1.0  0.15\n\
         \x20  SITE_CHAR  1  4  0.18  1.0  0.15"
            .to_string()
    };

    let ctrl = format!(
        "JOB\n\
         \x20  MESSAGES   {prefix}.msg\n\
         \x20  REPORT     {prefix}.rpt\n\
         \n\
         METPREP\n\
         \x20  DATA       {s2_prefix}_merge.out\n\
         \x20  XDATES     {year}/01/01  TO  {year}/12/31\n\
         \x20  OUTPUT     {sfc_out}\n\
         \x20  PROFILE    {pfl_out}\n\
         \x20  SURFDATA   {sf_station_id}  {year}\n\
         \x20  UAIRDATA   {ua_station_id}  {year}\n\
         \x20  SURFLOCN   {sf_lat:.4}  {sf_lon:.4}\n\
         \x20  UAIRLOCN   {ua_lat:.4}  {ua_lon:.4}\n\
         \x20  NWS_HGT    {anemometer_height}\n\
         \x20  METHOD     REFLEVEL  SUBNWS\n\
         {surf_chars}\n"
    );

    fs::write(&ctrl_path, ctrl).map_err(|e| format!("Write stage3: {e}"))?;
    Ok(ctrl_path)
}

// ---------------------------------------------------------------------------
// Execution
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize)]
pub struct AermetResult {
    pub success: bool,
    pub return_code: i32,
    #[serde(default)]
    pub messages: Vec<String>,
    #[serde(default)]
    pub warnings: Vec<String>,
    #[serde(default)]
    pub errors: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

pub fn run_aermet(
    ctrl_path: &Path,
    work_dir: &Path,
    progress_cb: &dyn Fn(&str, &str),
) -> AermetResult {
    let exe = match aermet_exe() {
        Some(e) => e,
        None => {
            return AermetResult {
                success: false,
                return_code: -1,
                messages: vec![],
                warnings: vec![],
                errors: vec!["AERMET executable not found".to_string()],
                error: Some(
                    "AERMET executable not found. Place aermet.exe in bin/".to_string(),
                ),
            }
        }
    };

    let ctrl_filename = ctrl_path
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();

    progress_cb("running", &format!("Running AERMET with {ctrl_filename}..."));

    let result = Command::new(&exe)
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .current_dir(work_dir)
        .spawn()
        .and_then(|mut child| {
            use std::io::Write;
            if let Some(ref mut stdin) = child.stdin {
                let _ = write!(stdin, "{ctrl_filename}\n");
            }
            child.wait_with_output()
        });

    match result {
        Ok(output) => {
            let msg_file = ctrl_filename.replace(".inp", ".msg");
            let msg_path = work_dir.join(&msg_file);
            let (messages, warnings, errors) = parse_message_file(&msg_path);

            let success = output.status.success() && errors.is_empty();
            let status = if success { "done" } else { "error" };
            let msg = if success { "OK" } else { "errors found" };
            progress_cb(status, &format!("AERMET {ctrl_filename}: {msg}"));

            AermetResult {
                success,
                return_code: output.status.code().unwrap_or(-1),
                messages,
                warnings,
                errors,
                error: None,
            }
        }
        Err(e) => AermetResult {
            success: false,
            return_code: -1,
            messages: vec![],
            warnings: vec![],
            errors: vec![format!("Failed to run AERMET: {e}")],
            error: Some(format!("Failed to run AERMET: {e}")),
        },
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct StageResults {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stage: Option<i32>,
    pub sfc_path: String,
    pub pfl_path: String,
    pub results: serde_json::Value,
}

pub fn run_aermet_all_stages(
    work_dir: &Path,
    year: i32,
    surface_file: &str,
    upper_air_file: &str,
    sf_station_id: &str,
    sf_lat: f64,
    sf_lon: f64,
    ua_station_id: &str,
    ua_lat: f64,
    ua_lon: f64,
    aersurface_file: Option<&str>,
    onemin_file: Option<&str>,
    anemometer_height: f64,
    manual_surface_chars: Option<&ManualSurfaceChars>,
    progress_cb: &dyn Fn(&str, &str),
) -> StageResults {
    let mut results = serde_json::Map::new();

    // Stage 1
    progress_cb("stage1", &format!("AERMET Stage 1 ({year}): Extract & QA..."));
    let ctrl1 = match generate_stage1(
        work_dir,
        year,
        surface_file,
        upper_air_file,
        sf_station_id,
        sf_lat,
        sf_lon,
        ua_station_id,
        ua_lat,
        ua_lon,
        onemin_file,
    ) {
        Ok(p) => p,
        Err(e) => {
            return StageResults {
                success: false,
                stage: Some(1),
                sfc_path: String::new(),
                pfl_path: String::new(),
                results: serde_json::json!({"error": e}),
            }
        }
    };

    let r1 = run_aermet(&ctrl1, work_dir, progress_cb);
    results.insert(
        "stage1".to_string(),
        serde_json::to_value(&r1).unwrap_or_default(),
    );
    if !r1.success {
        return StageResults {
            success: false,
            stage: Some(1),
            sfc_path: String::new(),
            pfl_path: String::new(),
            results: serde_json::Value::Object(results),
        };
    }

    // Stage 2
    progress_cb("stage2", &format!("AERMET Stage 2 ({year}): Merge..."));
    let ctrl2 = match generate_stage2(work_dir, year, onemin_file) {
        Ok(p) => p,
        Err(e) => {
            return StageResults {
                success: false,
                stage: Some(2),
                sfc_path: String::new(),
                pfl_path: String::new(),
                results: serde_json::json!({"error": e}),
            }
        }
    };

    let r2 = run_aermet(&ctrl2, work_dir, progress_cb);
    results.insert(
        "stage2".to_string(),
        serde_json::to_value(&r2).unwrap_or_default(),
    );
    if !r2.success {
        return StageResults {
            success: false,
            stage: Some(2),
            sfc_path: String::new(),
            pfl_path: String::new(),
            results: serde_json::Value::Object(results),
        };
    }

    // Stage 3
    progress_cb("stage3", &format!("AERMET Stage 3 ({year}): METPREP..."));
    let ctrl3 = match generate_stage3(
        work_dir,
        year,
        sf_station_id,
        sf_lat,
        sf_lon,
        ua_station_id,
        ua_lat,
        ua_lon,
        aersurface_file,
        anemometer_height,
        manual_surface_chars,
    ) {
        Ok(p) => p,
        Err(e) => {
            return StageResults {
                success: false,
                stage: Some(3),
                sfc_path: String::new(),
                pfl_path: String::new(),
                results: serde_json::json!({"error": e}),
            }
        }
    };

    let r3 = run_aermet(&ctrl3, work_dir, progress_cb);
    results.insert(
        "stage3".to_string(),
        serde_json::to_value(&r3).unwrap_or_default(),
    );

    let sfc_path = work_dir.join("output").join(format!("{year}.SFC"));
    let pfl_path = work_dir.join("output").join(format!("{year}.PFL"));

    StageResults {
        success: r3.success,
        stage: None,
        sfc_path: sfc_path.to_string_lossy().to_string(),
        pfl_path: pfl_path.to_string_lossy().to_string(),
        results: serde_json::Value::Object(results),
    }
}

// ---------------------------------------------------------------------------
// Message file parsing
// ---------------------------------------------------------------------------

fn parse_message_file(msg_path: &Path) -> (Vec<String>, Vec<String>, Vec<String>) {
    let mut messages = Vec::new();
    let mut warnings = Vec::new();
    let mut errors = Vec::new();

    if !msg_path.exists() {
        return (messages, warnings, errors);
    }

    let text = match fs::read_to_string(msg_path) {
        Ok(t) => t,
        Err(_) => return (messages, warnings, errors),
    };

    for line in text.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let upper = line.to_uppercase();
        if line.contains("***") && upper.contains("ERROR") {
            errors.push(line.to_string());
        } else if line.contains("***") && upper.contains("WARNING") {
            warnings.push(line.to_string());
        } else if line.starts_with("***") {
            messages.push(line.to_string());
        } else if line.starts_with("E ") || line.starts_with("E\t") {
            errors.push(line.to_string());
        } else if line.starts_with("W ") || line.starts_with("W\t") {
            warnings.push(line.to_string());
        }
    }

    (messages, warnings, errors)
}

/// Combine multiple per-year .SFC files into one.
pub fn combine_sfc_files(sfc_files: &[String], output_path: &Path) -> Result<(), String> {
    let mut out = fs::File::create(output_path).map_err(|e| format!("Create: {e}"))?;
    let mut sorted = sfc_files.to_vec();
    sorted.sort();
    for path in &sorted {
        let data = fs::read_to_string(path).map_err(|e| format!("Read {path}: {e}"))?;
        use std::io::Write;
        write!(out, "{data}").map_err(|e| format!("Write: {e}"))?;
    }
    Ok(())
}

/// Combine multiple per-year .PFL files into one.
pub fn combine_pfl_files(pfl_files: &[String], output_path: &Path) -> Result<(), String> {
    let mut out = fs::File::create(output_path).map_err(|e| format!("Create: {e}"))?;
    let mut sorted = pfl_files.to_vec();
    sorted.sort();
    for path in &sorted {
        let data = fs::read_to_string(path).map_err(|e| format!("Read {path}: {e}"))?;
        use std::io::Write;
        write!(out, "{data}").map_err(|e| format!("Write: {e}"))?;
    }
    Ok(())
}
